src/ast_editor.ml

```diff
@@ -0,0 +1,549 @@
+
+let put_keys_into_a_list data_serached map =
+  Map.filteri ~f:(fun ~key:_ ~data -> String.equal data_serached data) map
+  |> Map.to_alist |> List.unzip |> fst
+
+let find_webview_by_doc ~document =
+  match Map.find !webview_map (doc_string_uri ~document) with
+  | wv_opt when !original_mode -> wv_opt
+  | None -> (
+    match
+      put_keys_into_a_list (doc_string_uri ~document) !origin_to_pp_doc_map
+    with
+    | [ k ] -> (
+      match Map.find !webview_map k with
+      | wv_opt when not !original_mode -> wv_opt
+      | _ -> None)
+    | _ -> None)
+  | _ -> None
+
+let remove_keys_by_data map d =
+  map := Map.filteri ~f:(fun ~key:_ ~data -> not (String.equal data d)) !map
+
+let set_changes_tracking origin pp_doc =
+  origin_to_pp_doc_map :=
+    Map.set !origin_to_pp_doc_map
+      ~key:(doc_string_uri ~document:origin)
+      ~data:(doc_string_uri ~document:pp_doc);
+  pp_doc_to_changed_origin_map :=
+    Map.set
+      !pp_doc_to_changed_origin_map
+      ~key:(doc_string_uri ~document:pp_doc)
+      ~data:false
+
+let set_origin_changed ~data ~key =
+  pp_doc_to_changed_origin_map :=
+    Map.set !pp_doc_to_changed_origin_map ~key ~data
+
+let on_origin_update_content changed_document =
+  match
+    Map.find !origin_to_pp_doc_map (doc_string_uri ~document:changed_document)
+  with
+  | Some key -> set_origin_changed ~key ~data:true
+  | None -> ()
+
+let read_html_file () =
+  let filename = Node.__dirname () ^ "/../astexplorer/dist/index.html" in
+  Fs.readFile filename
+
+let document_eq a b =
+  String.equal
+    (Uri.toString (TextDocument.uri a) ())
+    (Uri.toString (TextDocument.uri b) ())
+
+let send_msg t value ~(webview : WebView.t) =
+  let msg = Ojs.empty_obj () in
+  Ojs.set_prop_ascii msg "type" (Ojs.string_to_js t);
+  Ojs.set_prop_ascii msg "value" value;
+  let _ = WebView.postMessage webview msg in
+  ()
+
+let transform_to_ast ~(document : TextDocument.t) ~(webview : WebView.t) =
+  let open Jsonoo.Encode in
+  let origin_json =
+    Dumpast.transform (TextDocument.getText document ()) (get_kind ~document)
+  in
+  let _pp_path = get_pp_path ~document in
+  let pp_value =
+    try
+      (*FIXME: adapt according to ppxlibs issue resulution *)
+      match get_preprocessed_ast (get_pp_path ~document) with
+      | Impl ppml_structure ->
+        let pp_code = get_reparsed_code_from_pp_file ~document in
+        let reparsed_structure =
+          pp_code |> Lexing.from_string |> Parse.implementation
+        in
+        let reparsed_json = Dumpast.reparse ppml_structure reparsed_structure in
+        reparsed_json
+      | Intf signature ->
+        let pp_code = get_reparsed_code_from_pp_file ~document in
+        let reparsed_signature =
+          pp_code |> Lexing.from_string |> Parse.interface
+        in
+        let reparsed_json =
+          Dumpast.reparse_signature signature reparsed_signature
+        in
+        reparsed_json
+    with
+    | _ -> null
```

Comment:
We should avoid doing a catch all on exceptions like this. Aren't there specific exceptions you can catch here?

Also, use the `match` to catche the exception:

```ocaml
match .. with
| Impl .. ->
| Intf .. ->
| exception _ -> null
```

Reply:
The only exception here was a `Failure` when reading binary file returned some kind of error. It has already been changed both in this function (d68aac57c09f3a65452aaa5ed6864eb1fad90ab3) and in [ppxlib#270](https://github.com/ocaml-ppx/ppxlib/pull/270). More changes are probably incoming though.